### ISSUE TITLE
[Router] `whileLoading` not displayed when using `<Set>`

### ISSUE THREAD

**Issue by dac09** [redwoodjs/redwood#2264]
### What?
When using `<Set>` with whileLoading props on routes, the whileLoading component is displayed for the first load (i.e. hard refresh) but not when navigating between routes.

### Repro
I have a repro of this live on the Tape site, but its a hard one to describe. My "whileLoading" component is the dashboard layout with the sheen bars

Routes.js looks like this:
```
     <Private unauthenticated="login">
        <Set wrap={DashboardLayout}>
          <Route
            path="/dashboard"
            page={DashboardHomePage}
            name="dashboard"
            whileLoading={DashboardLoader}
          />
```

Notice at 0:04, navigating from dashboard to profile, there's no sheen loaders in the "content" section
https://s.tape.sh/Ry7an6r2

When navigating between code split routes i.e. while its loading the JS chunk:
**Expected Behaviour**:
Should show `whileLoading` component

**Actual Behaviour**
Showing only the layout specified in Set's `wrap` prop

**Comment by dac09** [comment]
Tagging @Tobbe if you have any thoughts on this. I hope this makes sense!

**Comment by LBrian** [comment]
@dac09 The behaviour you described kinda makes sense for me purely based on the current implementation of `Set`. The `loading` of `useAuth` is the only state will trigger rendering the component assigned to `whileLoading`, since token is cached after the first load, so the following navigation will not trigger re-authentication thus no state changes of `loading`.

From architecture point of view, React is one-way data flow (mostly top-down i.e. `parent -> children`), so I can't think of a clean way (or necessity) to access children's state then render `whileLoading` accordingly, it sounds for me a bit violating its architecture design.

I ran into the similar situation at the moment wanting to render a transition page during authentication, looked into the code base and rejustified the current logic, I suggest (believe) that we should move the prop `whileLoading` to `Set` and deprecate it from `Route`

- `whileLoading` depends on `useAuth` only
- we don't detect `Page` state in `PageLoader`
- might be tricky trying to do ☝️ 
- a bit redundant if I want to assign the same transition page for all private pages (e.g. 👇 )

![image](https://user-images.githubusercontent.com/3888780/122153928-46588a00-cea7-11eb-84b9-4e5d8f04013c.png)
instead I can just do this
![image](https://user-images.githubusercontent.com/3888780/122153992-61c39500-cea7-11eb-9f26-c316f1830b1e.png)

Hopefully this makes sense to you and I will post this in Discord to get more inputs and I am happy to create a PR for this later if we all agreed this is the right direction

**Comment by Tobbe** [comment]
Thanks for bringing this up. This is a good discussion to have.

I think we actually want both. You should be able to set `whileLoading` both on individual routes and on an entire `<Set>`. And it should not only work for waiting for auth. It should also be for when we're waiting for the JS for a page to download (downloading due to code splitting).

**Comment by LBrian** [comment]
> It should also be for when we're waiting for the JS for a page to download (downloading due to code splitting).
This is not in place yet right? If this is TODO then make sense to have both

**Comment by Tobbe** [comment]
> > It should also be for when we're waiting for the JS for a page to download (downloading due to code splitting).
> 
> This logic is not in place yet right? I did thought about having both after left the comment, good callout.

I was remembering wrong. We handle page downloading in a different way. https://redwoodjs.com/docs/redwood-router#pageloadingcontext

`whileLoading` is only for auth, just like you said

**Comment by dac09** [comment]
> > > It should also be for when we're waiting for the JS for a page to download (downloading due to code splitting).
> > 
> > 
> > This logic is not in place yet right? I did thought about having both after left the comment, good callout.
> 
> I was remembering wrong. We handle page downloading in a different way. https://redwoodjs.com/docs/redwood-router#pageloadingcontext
> 
> `whileLoading` is only for auth, just like you said

Confirming this behaviour. I'm fairly sure we used to have `whileLoading` also trigger during chunk loading. 🤷

I haven't looked at the internals of how this works (because Router...) - but I do think having it also be triggered while the page chunk is downloading would be great. I think its certainly more ergonomic than manually creating a loader with pageLoadingContext, don't you think? 

I'm not at all dogmatic about this, and if you think this isn't the right approach, can close this issue

**Comment by LBrian** [comment]
@dac09 I do agree with you the benefit of having `whileLoading` in `Route`, quickly checked the code and plausibly speaking we should be able to achieve by passing `whileLoading` down to `PageLoader`

![image](https://user-images.githubusercontent.com/3888780/122318992-5170f000-cf63-11eb-994d-b3b5e1cdfb86.png)

I'm not 100% sure for ☝️, it would be good to bring in more thoughts. 

However, I do suggest we should add `whileLoading` to `<Set>` and `<Private>` as priority to align with the current logic and avoid confusing developers, also saw @Tobbe commented on this. I can create a PR for this if we all agree on this.

![image](https://user-images.githubusercontent.com/3888780/122319549-17ecb480-cf64-11eb-8fa0-6cf729868390.png)

**Comment by Tobbe** [comment]
@peterp mentioned maybe using a React Suspense boundary for this. Is that something you're familiar with @LBrian? Do you think it could work?

**Comment by LBrian** [comment]
> @peterp mentioned maybe using a React Suspense boundary for this. Is that something you're familiar with @LBrian? Do you think it could work?

@Tobbe Good callout! We should use `Suspense` under the hood for this

**Comment by Tobbe** [comment]
Great! Do you want to try to update this PR to use Suspense?

**Comment by LBrian** [comment]
> Great! Do you want to try to update this PR to use Suspense?

Just assign to me and I will take care of it.

**Comment by dac09** [comment]
Suspense is totally a good call for this - assigning to you @LBrian - thank you!

**Comment by LBrian** [comment]
@Tobbe @peterp I know we agreed to use `Suspense` but it will become breaking changes from my understanding, see my verdict in #2889. I believe this is an easy and quick fix without changing current `PageLoader` and `usePageLoadingContext` architecture

**Comment by dac09** [comment]
Fixed by #2889 🎉
